### PR TITLE
Optimize JUMPDEST analysis

### PR DESCRIPTION
**Pull request by chfast**
Optimize the build_jumpdest_map() procedure for JUMPDEST analysis.
- "is push opcode" conditions has been changed to (op & 0xE0) == 0x60.
- the op == JUMPDEST condition has been marked as unlikely.

### CODE REVIEW

**Review by holiman — APPROVED** [review]
lgtm

**Comment by chfast** [comment]
### Benchmarks

Time are in µs.

Haswell 4.0 GHz, Clang 12
```
Comparing o/ba-0 to o/ba-opt                                                                                                                                                                                                      
Benchmark                                                                Time             CPU      Time Old      Time New       CPU Old       CPU New                                                                             
-----------------------------------------------------------------------------------------------------------------------------------------------------
baseline/analyse/main/blake2b_huff_mean                               -0.6025         -0.6025            15             6            15             6
baseline/analyse/main/blake2b_shifts_mean                             -0.6836         -0.6836            11             4            11             4
baseline/analyse/main/sha1_divs_mean                                  -0.7625         -0.7625             2             1             2             1
baseline/analyse/main/sha1_shifts_mean                                -0.7681         -0.7681             2             1             2             1
baseline/analyse/main/weierstrudel_mean                               -0.5738         -0.5738            18             8            18             8
baseline/analyse/micro/beginsub_push1s_0xffff_mean                    +0.0411         +0.0411            62            65            62            65
baseline/analyse/micro/beginsubs_0xffff_mean                          +0.3286         +0.3286            25            33            25            33
baseline/analyse/micro/jumpdests_0xffff_mean                          +0.0700         +0.0700            80            86            80            86
baseline/analyse/micro/loop_with_many_jumpdests_mean                  +0.0730         +0.0730            30            32            30            32
baseline/analyse/micro/push1s_0xffff_mean                             +0.0363         +0.0363            66            68            66            68
baseline/analyse/micro/push32s_0xffff_mean                            +0.0407         +0.0407             4             4             4             4
baseline/analyse/micro/zeros_0xffff_mean                              +0.3301         +0.3301            25            33            25            33
```

Haswell 4.0 GHz, GCC 10
```
Comparing o/ba-0 to o/ba-opt                                                                                                                                                                                                      
Benchmark                                                                Time             CPU      Time Old      Time New       CPU Old       CPU New                                                                             
-----------------------------------------------------------------------------------------------------------------------------------------------------
baseline/analyse/main/blake2b_huff_mean                               -0.2021         -0.2021             8             6             8             6
baseline/analyse/main/blake2b_shifts_mean                             -0.0415         -0.0415             4             4             4             4
baseline/analyse/main/sha1_divs_mean                                  -0.0608         -0.0608             1             1             1             1
baseline/analyse/main/sha1_shifts_mean                                -0.0591         -0.0591             1             1             1             1
baseline/analyse/main/weierstrudel_mean                               -0.0665         -0.0665             9             8             9             8
baseline/analyse/micro/beginsub_push1s_0xffff_mean                    -0.0131         -0.0131            71            70            71            70
baseline/analyse/micro/beginsubs_0xffff_mean                          +0.0005         +0.0005            33            33            33            33
baseline/analyse/micro/jumpdests_0xffff_mean                          +0.0153         +0.0153            89            90            89            90
baseline/analyse/micro/loop_with_many_jumpdests_mean                  +0.0171         +0.0171            33            34            33            34
baseline/analyse/micro/push1s_0xffff_mean                             -0.0083         -0.0083            75            74            75            74
baseline/analyse/micro/push32s_0xffff_mean                            -0.0059         -0.0059             5             5             5             5
baseline/analyse/micro/zeros_0xffff_mean                              -0.0006         -0.0006            33            33            33            33
```

**Review by yperbasis — APPROVED** [review]

File: lib/evmone/baseline.cpp [hunk 1]
```diff
@@ -12,16 +12,21 @@ namespace evmone
...
+
+    JumpdestMap map(code_size);
     for (size_t i = 0; i < code_size; ++i)
     {
         const auto op = code[i];
-        if (op == OP_JUMPDEST)
-            m[i] = true;
-        else if (op >= OP_PUSH1 && op <= OP_PUSH32)
-            i += op - size_t{OP_PUSH1 - 1};
+        if ((op & push_op_mask) == OP_PUSH1)
```
Comment (gumb0) [inline lib/evmone/baseline.cpp:24]: Nit: maybe simpler
```suggestion
        if ((op & push_op_mask) == push_op_mask)
```
or add `static_assert(push_op_mask == OP_PUSH1)`

**Review by gumb0 — APPROVED** [review]

File: lib/evmone/baseline.cpp (hunk 1 above)
Comment (chfast) [reply lib/evmone/baseline.cpp:24]: These are different. The mask is 0xE0.

File: lib/evmone/baseline.cpp [hunk 2]
```diff
@@ -12,16 +12,21 @@ namespace evmone
 {
 JumpdestMap build_jumpdest_map(const uint8_t* code, size_t code_size)
 {
-    JumpdestMap m(code_size);
+    // Bitmask for PUSH instructions identification.
+    // It removes the lower 5 bits of a PUSH instruction with information about the data length.
+    // The remaining top 3 bits of every PUSH instruction is 0b011, i.e. 0x60, OP_PUSH1.
```
Comment (gumb0) [inline lib/evmone/baseline.cpp:17]: This sentence confused me, but probably it's just me.

Comment (chfast) [reply lib/evmone/baseline.cpp]: Better now?

**Pull request merged**